HW2/P5/physics.pyx

```diff
@@ -55,76 +59,153 @@ cdef void sub_update(FLOAT[:, ::1] XY,
+                     omp_lock_t *locks) nogil:
     cdef:
         FLOAT *XY1, *XY2, *V1, *V2
         int j, dim
         float eps = 1e-5
 
     # SUBPROBLEM 4: Add locking
+    # GET THE LOCK HERE SO THAT THE TWO THREADS DO NOT ATTEMPT TO WRITE WITHIN THE 
+    # SAME MEMORY LOCATION. GIVEN THE THREADING SCHEME, DEADLOCKING CAN BE AVOIDED BY
+    # THE COMPARISON DONE BELOW
+    acquire(&(locks[i]))
+
     XY1 = &(XY[i, 0])
     V1 = &(V[i, 0])
+    
     #############################################################
     # IMPORTANT: do not collide two balls twice.
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
 
+    cdef int xCenter = <int>(XY[i, 0]/grid_spacing)
+    cdef int yCenter = <int>(XY[i, 1]/grid_spacing)
+    
+    # OBTAIN THE BALLS THAT ARE WITHIN THE GRIDS DEFINED ABOVE
+    # THE GRID SPACE IS SUPPOSED TO BE MUCH SMALLER THAN HAVING
+    # TO LOOK AT EACH BALL EVERYWERE FOR EVERY COLLISION
+    cdef int distCheck = 3
+    cdef int xMaxLim = xCenter+distCheck
+    cdef int xMinLim = xCenter-distCheck
+    cdef int yMaxLim = yCenter+distCheck
+    cdef int yMinLim = yCenter-distCheck
+
+    cdef int gridXMax = Grid.shape[0]-1
+    cdef int gridYMax = Grid.shape[1]-1
+
+    # CHECK TO SEE IF THE GRID BOUNDARIES EXIT THE DOMAIN BOUNDARIES, 
+    # IF THEY DO, BRING THE GRID BOUNDARIES BACK INTO THE DOMAIN
+    if xMaxLim>=gridXMax: xMaxLim = gridXMax
+    if xMinLim<0: xMinLim=0
+    if yMaxLim>=gridYMax: yMaxLim = gridYMax
+    if yMinLim<0: yMinLim=0
+    
+    cdef int ball_grids
+    cdef int xx, yy
+    
+    # NOW INSTEAD OF CHECKING BETWEEEN EACH PARTICLE IN THE ENTIRE DOMAIN
+    # CUT DOWN ON ONLY THE PARTICLES THAT ARE IN THE GRIDS NEXT TO THE 
+    # CURRENT GRID OF INTEREST. THIS REDUCES THE COMPUTATION FROM ORDER N^2 TO 
+    # ORDER N
+    for xx in range(xMinLim,xMaxLim+1):
+        for yy in range(yMinLim,yMaxLim+1):
+            ball_grids= <int> Grid[xx,yy]
+            # THE STATEMENT ON THE RIGHT AVOIDS DEADLOCKING AND ENSURES THAT BALLS 
+            # ARE DO NOT ACCOUNT FOR COLIDING TWICE, AS IT ONLY ALLOWS IT TO CONTINUE
+            # AT A CERTIAN ORDER.
+            if ball_grids != negative_one and ball_grids>i: 
+                XY2 = &(XY[ball_grids, 0])
+                V2 = &(V[ball_grids, 0])
+                if overlapping(XY1, XY2, R):
+                    # SUBPROBLEM 4: Add locking
+                    if not moving_apart(XY1, V1, XY2, V2) and XY2!=XY1:
```

Comment:
Using your locking scheme, is it possible for two balls that are determined to be moving apart as determined by the above comparison to no longer be moving apart by the time the collide function is called?